xray/test/test_plot.py

```diff
@@ -165,18 +167,136 @@ def test_plot_nans(self):
 class TestDetermineCmapParams(TestCase):
     def test_robust(self):
         data = np.random.RandomState(1).rand(100)
-        vmin, vmax, cmap, extend = _determine_cmap_params(data, robust=True)
+        vmin, vmax, cmap, extend, levels, cnorm = _determine_cmap_params(
+            data, robust=True)
         self.assertEqual(vmin, np.percentile(data, 2))
         self.assertEqual(vmax, np.percentile(data, 98))
         self.assertEqual(cmap.name, 'viridis')
         self.assertEqual(extend, 'both')
+        self.assertIsNone(levels)
+        self.assertIsNone(cnorm)
 
     def test_center(self):
         data = np.random.RandomState(2).rand(100)
-        vmin, vmax, cmap, extend = _determine_cmap_params(data, center=0.5)
+        vmin, vmax, cmap, extend, levels, cnorm = _determine_cmap_params(
+            data, center=0.5)
         self.assertEqual(vmax - 0.5, 0.5 - vmin)
         self.assertEqual(cmap, 'RdBu_r')
         self.assertEqual(extend, 'neither')
+        self.assertIsNone(levels)
+        self.assertIsNone(cnorm)
+
+    def test_integer_levels(self):
+        data = 1 + np.random.RandomState(3).rand(100)
+        vmin, vmax, cmap, extend, levels, cnorm = _determine_cmap_params(
+            data, levels=5, vmin=0, vmax=5, cmap='Blues')
+        self.assertEqual(vmin, levels[0])
+        self.assertEqual(vmax, levels[-1])
+        self.assertEqual(cmap.name, 'Blues')
+        self.assertEqual(extend, 'neither')
+        self.assertEqual(cmap.N, 5)
+        self.assertEqual(cnorm.N, 6)
+
+        vmin, vmax, cmap, extend, levels, cnorm = _determine_cmap_params(
+            data, levels=5, vmin=0.5, vmax=1.5)
+        self.assertEqual(cmap.name, 'viridis')
+        self.assertEqual(extend, 'max')
+
+    def test_list_levels(self):
+        data = 1 + np.random.RandomState(3).rand(100)
+
+        orig_levels = [0, 1, 2, 3, 4, 5]
+        # vmin and vmax should be ignored if levels are explicitly provided
+        vmin, vmax, cmap, extend, levels, cnorm = _determine_cmap_params(
```

Comment:
Another possible implementation to consider is having `_determine_cmap_params` return `locals()`. Then we can use a dictionary rather than doing tuple unpacking many times. Labeled data FTW :)